pennylane/interfaces/autograd.py

```diff
@@ -195,18 +195,81 @@ def vjp(ans, self, params, device):  # pylint: disable=unused-argument
-            # pass, so we do not need to re-unwrap the parameters.
+        saved_grad_matrices = {}
+
+        def _evaluate_grad_matrix(p, grad_matrix_fn):
+            """Convenience function for generating gradient matrices
+            for the given parameter values.
+
+            This function serves two purposes:
+
+            * Avoids duplicating logic surrounding parameter unwrapping/wrapping
+
+            * Takes advantage of closure, to cache computed gradient matrices via
+              the ``saved_grad_matrices`` attribute, to avoid gradient matrices being
+              computed multiple redundant times.
+
+              This is particularly useful when differentiating vector-valued QNodes.
+              Because Autograd requests the vector-grad matrix product,
+              and *not* the full grad matrix, differentiating vector-valued
+              functions will result in multiple backward passes.
+
+            Args:
+                p (Sequence): quantum tape parameter values use to evaluate the gradient matrix
+                grad_matrix_fn (str): Name of the gradient matrix function. Should correspond to an existing
+                    tape method. Currently allowed values include ``"jacobian"`` and ``"hessian"``.
+
+                Returns:
+                    array[float]: the gradient matrix
+            """
+            if grad_matrix_fn in saved_grad_matrices:
+                return saved_grad_matrices[grad_matrix_fn]
+
             self.set_parameters(self._all_params_unwrapped, trainable_only=False)
-            jac = self.jacobian(device, params=params, **self.jacobian_options)
+            grad_matrix = getattr(self, grad_matrix_fn)(device, params=p, **self.jacobian_options)
             self.set_parameters(self._all_parameter_values, trainable_only=False)
 
-            # only flatten g if all parameters are single values
+            saved_grad_matrices[grad_matrix_fn] = grad_matrix
+            return grad_matrix
+
+        def gradient_product(dy):
+            """Returns the vector-Jacobian product with given
+            parameter values p and output gradient dy"""
+
+            @autograd.extend.primitive
+            def jacobian(p):
+                """Returns the Jacobian for parameters p"""
+                return _evaluate_grad_matrix(p, "jacobian")
+
+            def vhp(ans, p):
+                def hessian_product(ddy):
+                    """Returns the vector-Hessian product with given
+                    parameter values p, output gradient dy, and output
+                    second-order gradient ddy"""
+                    hessian = _evaluate_grad_matrix(p, "hessian")
+
+                    if dy.size > 1:
+                        if all(np.ndim(p) == 0 for p in params):
+                            # only flatten dy if all parameters are single values
+                            vhp = dy.flatten() @ ddy @ hessian @ dy.flatten()

```

Comment:
I don't understand why we have `dy` twice here.

Reply:
I'm not sure I can fully answer that, but it makes sense if you consider the shape of the hessian and perform dimensional analysis - the hessian will be of shape `(num_params, num_params, num_output)`, while `dy` has shape `num_output`, and `ddy` has shape `(num_output, num_params)`.

So performing the matrix multiplication will return a vhp of the right size - `(num_params,)`.

E.g.:

```python
>>> num_params = 2
>>> num_output = 3
>>> dy = np.random.random([num_output])
>>> ddy = np.random.random([num_output, num_params])
>>> hessian = np.random.random([num_params, num_params, num_output])
>>> dy @ ddy @ hessian @ dy
array([1.82038363, 2.03458631])
```